### PR TITLE
Update database connection code

### PR DESCRIPTION
**Pull request by mattias-p**
## Purpose

This PR updates the database connection code so that it's possible to get a database adapter instance that doesn't correspond to the contents of a configuration file. This is useful if you want to include database schema definition in the database adapter code.

This is mostly about refactoring, but it also includes fixing of a bug (introduced since last release) and some minor improvements.

## Context

This PR is a step towards both #805 and #806.

## Changes

* Fixes connecting to MySQL over IPv6.

* Fixes the syntax of the data source name for PostgreSQL. I'm not sure why the old syntax used to work or why it doesn't anymore (on CentOS 8), but this PR changes the syntax to the one specified in [DBD::Pg::connect](https://metacpan.org/pod/DBD::Pg#connect).

* The database adapter constructor interfaces are changed from `new({ config => $config })`to `from_config( $config )`. This makes room for adding other constructors that don't take their parameters straight out of a Config instance.

* There are a number of methods in the database adapters that used to look into the configuration for parameters. Now they are updated to take those parameters as arguments instead.

* Database adapters used to defer the database connection for as long as possible. However in order to avoid failures in inconvenient places we were explicitly working around that feature by forcing the opening of a database connection as early as possible. This PR removes the connection deferral feature.

* The DRY principle is applied to getting the database adapter class and to opening database connections. The consistency across adapters is improved regarding log messages and database connection settings, and it also helps us stay consistent.

## How to test this PR

1. Verify that Travis was able to run the tests using all database engines.
2. For each database engine:
    1. Configure Backend to use this database engine
    2. Restart zm-rpcapi and zm-testagent
    3. Run a test using zmtest and verify you get a result back.

### CODE REVIEW

File: lib/Zonemaster/Backend/DB.pm [hunk 1]
```diff
@@ -68,14 +68,14 @@ sub add_api_user {
 
 # Standard SQL, can be here
 sub get_test_request {
-    my ( $self ) = @_;
+    my ( $self, $queue ) = @_;
```
Comment (pnax) [inline lib/Zonemaster/Backend/DB.pm:72]: Why not naming it `$lock_on_queue` directly ?

**Review by pnax — COMMENTED** [review]
This is a pretty nice work. A remark though, you choose to use the variable name `$queue`, I feel like this is not really expressive. Usually a queue is a set of items. Here it is just holding the value of the `lock_on_queue` parameter. Do you mind using a more explicit term? If you dislike the "lock_on_queue" term, it might be something like "queue_label" or "queue_value".

File: lib/Zonemaster/Backend/DB.pm (hunk 1 above)
Comment (mattias-p) [reply lib/Zonemaster/Backend/DB.pm:72]: Good point! I'll go for `$queue_label`.

**Review by matsduf — CHANGES_REQUESTED** [review]
I created an installation package based on this PR and tried to install it on a VPS with no MySQL installed.

```
# cpanm -iv Zonemaster-Backend-6.2.0-PR-4499204-2021-07-02-B.tar.gz
(...)
t/test_validate_syntax.t .. 1/2 Internal error 002 

#   Failed test 'no warnings'
#   at /usr/local/lib/perl5/5.32/Test/Builder.pm line 153.
# There were 1 warning(s)
# 	Previous test 1 'use Zonemaster::Backend::RPCAPI;'
# 	Internal error 002: Unexpected error in the _init_db API call: [Failed to initialize the [MySQL] database backend module: [install_driver(mysql) failed: Can't locate DBD/mysql.pm in @INC (you may need to install the DBD::mysql module) (@INC contains: /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/inc /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/blib/lib /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/blib/arch /usr/local/lib/perl5/site_perl/mach/5.32 /usr/local/lib/perl5/site_perl /usr/local/lib/perl5/5.32/mach /usr/local/lib/perl5/5.32 .) at (eval 458) line 3. Perhaps the DBD::mysql perl module hasn't been fully installed, or perhaps the capitalisation of 'mysql' isn't right. Available drivers: DBM, ExampleP, File, Gofer, Mem, Pg, Proxy, SQLite, Sponge.  at /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/blib/lib/Zonemaster/Backend/DB.pm line 167. ]] 
#  at /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/blib/lib/Zonemaster/Backend/RPCAPI.pm line 78.
# 	Zonemaster::Backend::RPCAPI::handle_exception("_init_db", "Failed to initialize the [MySQL] database backend module: [in"..., 002) called at /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/blib/lib/Zonemaster/Backend/RPCAPI.pm line 68
# 	Zonemaster::Backend::RPCAPI::_init_db(Zonemaster::Backend::RPCAPI=HASH(0x80662a498), "MySQL") called at /usr/home/matsd/.cpanm/work/1625225453.1068/Zonemaster-Backend-6.2.0/blib/lib/Zonemaster/Backend/RPCAPI.pm line 56
# 	Zonemaster::Backend::RPCAPI::new("Zonemaster::Backend::RPCAPI", HASH(0x806570a38)) called at t/test_validate_syntax.t line 30
# 	main::__ANON__() called at /usr/local/lib/perl5/5.32/Test/Builder.pm line 334
# 	eval {...} called at /usr/local/lib/perl5/5.32/Test/Builder.pm line 334
# 	Test::Builder::subtest(Test::Builder=HASH(0x800bc5630), "Everything but NoWarnings", CODE(0x800b1bc30)) called at /usr/local/lib/perl5/5.32/Test/More.pm line 809
# 	Test::More::subtest("Everything but NoWarnings", CODE(0x800b1bc30)) called at t/test_validate_syntax.t line 243
# 
# Looks like your test exited with 255 just after 2.
(...)
```

File: lib/Zonemaster/Backend/DB/MySQL.pm [hunk 2]
```diff
@@ -346,9 +376,9 @@ sub add_batch_job {
 }
 
 sub select_unfinished_tests {
-    my ( $self ) = @_;
+    my ( $self, $queue, $test_run_timeout, $test_run_max_retries ) = @_;
```
Comment (pnax) [inline lib/Zonemaster/Backend/DB/MySQL.pm:379]: here as well

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm [hunk 3]
```diff
@@ -318,9 +349,9 @@ sub add_batch_job {
 }
 
 sub select_unfinished_tests {
-    my ( $self ) = @_;
+    my ( $self, $queue, $test_run_timeout, $test_run_max_retries ) = @_;
```
Comment (pnax) [inline lib/Zonemaster/Backend/DB/PostgreSQL.pm:352]: same

File: lib/Zonemaster/Backend/DB.pm [hunk 4]
```diff
@@ -102,12 +120,16 @@ sub get_batch_job_result {
 }
 
 sub process_unfinished_tests {
-    my ( $self ) = @_;
+    my ( $self, $queue, $test_run_timeout, $test_run_max_retries ) = @_;
```
Comment (pnax) [inline lib/Zonemaster/Backend/DB.pm:123]: Could you do the same update on the "queue" variable name here as well?

File: lib/Zonemaster/Backend/DB/SQLite.pm [hunk 5]
```diff
@@ -373,9 +369,9 @@ sub add_batch_job {
 }
 
 sub select_unfinished_tests {
-    my ( $self ) = @_;
+    my ( $self, $queue, $test_run_timeout, $test_run_max_retries ) = @_;
```
Comment (pnax) [inline lib/Zonemaster/Backend/DB/SQLite.pm:372]: same

**Comment by mattias-p** [comment]
@matsduf Good spotting! This part of the validation code depends on having a database adapter instance, and by default the database engine specified in the default config file (MySQL) was used. There's no good reason for this dependency, but rather it's a design flaw.

The test used to work because the adapters connected lazily, and since validate_syntax() never uses the database no connection attempt would be made. With this PR all adapters (including MySQL) connect eagerly.

I resolved this by hardcoding test_validate_syntax.t to use SQLite, which is the default for test01.t. A more proper fix would be to make the validation code usable without any database adapter at all, but I feel that's out of scope for this PR.

**Comment by mattias-p** [comment]
@pnax and @matsduf Please review again.

**Review by pnax — COMMENTED** [review]
Could you update to `$queue_label` in other places as well?

File: lib/Zonemaster/Backend/DB/MySQL.pm (hunk 2 above)
Comment (mattias-p) [reply lib/Zonemaster/Backend/DB/MySQL.pm]: I've replaced all occurrances of `$queue` with `$queue_label`.

**Review by matsduf — APPROVED** [review]

**Review by pnax — APPROVED** [review]

**Pull request merged**